### PR TITLE
[RLlib] DDPG refactor and Exploration API action noise classes.

### PR DESCRIPTION
**Pull request by sven1977**
- The Exploration API is missing to complete P0 tasks:
GaussianNoise and UrnsteinUhlenbeckNoise Exploration classes. These two are added with this PR (plus test cases).
- DDPG (and TD3) are refactored to use these two classes.

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://ray.readthedocs.io/en/latest/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.

### CODE REVIEW

File: rllib/tests/test_checkpoint_restore.py [hunk 1]
```diff
@@ -144,14 +146,14 @@ def valid_tf_checkpoint(checkpoint_dir):
...
-        for name in [
-                "SAC", "ES", "DQN", "DDPG", "PPO", "A3C", "APEX_DDPG", "ARS"
-        ]:
+        for name in ["DDPG"]:
             test_ckpt_restore(use_object_store, name, failures)
 
     assert not failures, failures
     print("All checkpoint restore tests passed!")
 
+    quit()
```
Comment (ericl) [inline rllib/tests/test_checkpoint_restore.py]: Revert the changes in this file for merge?

File: rllib/utils/exploration/gaussian_noise.py [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+        else:
+            action = action_dist.deterministic_sample()
+
+        # Logp=always zero.
+        logp = torch.zeros(shape=(action.size()[0], ), dtype=torch.float32)
+
+        return action, logp
+
+    @override(Exploration)
+    def get_info(self):
```
Comment (ericl) [inline rllib/utils/exploration/gaussian_noise.py]: I think it makes more sense for this to be a separate `get_scale_value` method rather than overload the existing info method (which seems like it's only for debugging).

**Review by ericl — COMMENTED** [review]
What's the testing strategy here, should we rerun a few of the mujoco benchmarks?

File: rllib/tests/test_checkpoint_restore.py (hunk 1 above)
Comment (sven1977) [reply rllib/tests/test_checkpoint_restore.py]: Ups, sorry, thanks for catching this!

Comment (sven1977) [reply rllib/tests/test_checkpoint_restore.py]: Done.

File: rllib/utils/exploration/gaussian_noise.py (hunk 2 above)
Comment (sven1977) [reply rllib/utils/exploration/gaussian_noise.py]: Not sure I agree. Yes, right now, it's only used for recording this "info" (whatever it is for each Exploration class) in the train-results (Policy calls this generically and then tune json's it). This is useful, just like it is for current learning rate, memory consumption, etc.. (debugging/reporting snapshot of the state of the exploration object). I don't see any use for a specific `get_scale_value` method, as it would never be used by anyone (Policy should only use the top-level Exploration class methods, as it doesn't know, which specific sub-class it's holding).

**Comment by sven1977** [comment]
Yes, I'll rerun HalfCheetah vs DDPG. ...

File: rllib/utils/exploration/gaussian_noise.py (hunk 2 above)
Comment (ericl) [reply rllib/utils/exploration/gaussian_noise.py]: I was referring to the comment you had here:
```
            clean_actions, cur_noise_scale = self.sess.run(
                [self.output_actions,
                 self.exploration.get_info()],
```

It seemed odd to be passing self.exploration.get_info() to the session run, but if this is a temporary hack I guess it's fine. Otherwise, it seems more clear to assert the exploration is of type GaussianNoise, and then call a gaussian-noise specific method, rather than sneak the tensor through via get_info().

**Review by ericl — COMMENTED** [review]
Looks good to me, pending verification of DDPG/TD3 results. We should probably add a couple MuJoCo regression tests to our standard compact-regression-test suite, it will be a little annoying because of the licensing.

Any suggestions on alternative OSS continuous benchmarks?

**Review by ericl — APPROVED** [review]

File: rllib/utils/exploration/gaussian_noise.py (hunk 2 above)
Comment (sven1977) [reply rllib/utils/exploration/gaussian_noise.py]: Ah yes, absolutely. This is only a problem though right now as we don't have a parameter-noise exploration class yet and the parameter-noise logic piggybacks on that cur_noise_scale value (which it shouldn't, it should be separate exploration classes or a merged one). It's tagged as a `TODO(sven)` and will be fixed once we add ParameterNoise exploration classes.

**Comment by sven1977** [comment]
@ericl Please merge. DDPG looks still good. I'll run a longer benchmark vs HalfCheetah and update the rl-experiments repo.

```
== Status ==
Memory usage on this node: 7.9/120.0 GiB
Using FIFO scheduling algorithm.
Resources requested: 0/16 CPUs, 0/1 GPUs, 0.0/88.48 GiB heap, 0.0/12.84 GiB objects
Result logdir: /home/ubuntu/ray_results/halfcheetah-ddpg
Number of trials: 1 (1 TERMINATED)
+---------------------------+------------+-------+----------+------------------+--------+--------+
| Trial name                | status     | loc   |   reward |   total time (s) |     ts |   iter |
|---------------------------+------------+-------+----------+------------------+--------+--------|
| DDPG_HalfCheetah-v2_00000 | TERMINATED |       |  2006.98 |          3003.82 | 211000 |    211 |
+---------------------------+------------+-------+----------+------------------+--------+--------+
```
